src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,534 @@
+      lhs.set_data(input_A_prepared);
+
+      // PRINT_MATRIX_DEBUG(input_A_prepared, rows_A, width, Order::RowMajor);
+
+      ruy::Matrix<std::int8_t> rhs;
+      ruy::MakeSimpleLayout(width, cols_B, ruy::Order::kColMajor, rhs.mutable_layout());
+      rhs.set_data(input_B_prepared);
+
+      // PRINT_MATRIX_DEBUG(input_B_prepared, width, cols_B, Order::ColMajor);
+
+      ruy::Matrix<std::int32_t> dst;
+      ruy::MakeSimpleLayout(rows_A, cols_B, ruy::Order::kRowMajor, dst.mutable_layout());
+
+      AlignedVector<std::int32_t> dst_data(rows_A * cols_B);
+      std::int32_t *dest_ptr = dst_data.data();
+
+      dst.set_data(dest_ptr);
+
+      // When Dst is int32, mul_params is unused.
+      ruy::MulParams<std::int32_t, std::int32_t> mul_params;
+      ruy::Mul(lhs, rhs, mul_params, &context, &dst);
+
+      // Unquantizes, then adds bias in a single statement on the output.
+      // float unquant_multiplier = (1.0f * scale_output) / (scale_A * scale_B);
+      Preprocess<kHighestPath>::unquantizeAddBias(
+          dest_ptr, bias_prepared, unquant_multiplier, rows_A, cols_B, output);
+    }
+
+    // output = A*B (notice no bias).
+    static void Multiply(const Type *input_A_prepared,
+                         const Type *input_B_prepared,
+                         float *output,
+                         Index rows_A,
+                         Index width,
+                         Index cols_B,
+                         float unquant_multiplier) {
+      // It is expected that somehow we have managed to call all prepare by the time
+      // we are here, with inputs (prepared) in int8_t. All that's left to do is use
+      // ruy for multiply and then start with the reverse ops to get to fp32.
+
+      // Use ruy to multiply.
+      // The following is adapted from
+      // https://github.com/google/ruy/blob/878283640de7946a43053e8ebf4f15114fbc9156/example/example.cc#L129-L152
+
+      ruy::Context context;
+      ruy::Matrix<std::int8_t> lhs;
+      ruy::MakeSimpleLayout(rows_A, width, ruy::Order::kRowMajor, lhs.mutable_layout());
+      lhs.set_data(input_A_prepared);
+
+      ruy::Matrix<std::int8_t> rhs;
+      ruy::MakeSimpleLayout(width, cols_B, ruy::Order::kColMajor, rhs.mutable_layout());
+      rhs.set_data(input_B_prepared);
+
+      ruy::Matrix<std::int32_t> dst;
+      ruy::MakeSimpleLayout(rows_A, cols_B, ruy::Order::kRowMajor, dst.mutable_layout());
+
+      AlignedVector<std::int32_t> dst_data(rows_A * cols_B);
+      std::int32_t *dest_ptr = dst_data.data();
+
+      dst.set_data(dest_ptr);
+
+      // When Dst is int32, mul_params is unused.
+      ruy::MulParams<std::int32_t, std::int32_t> mul_params;
+      ruy::Mul(lhs, rhs, mul_params, &context, &dst);
+
+      // Unquantizes, then adds bias in a single statement on the output.
+      // float unquant_multiplier = (1.0f * scale_output) / (scale_A * scale_B);
+      Preprocess<kHighestPath>::unquantize(dest_ptr, unquant_multiplier, rows_A, cols_B, output);
+    }
```

Comment:
y not, `if (bias) { unquantiseAddBias } else { unquantise}` ?

Reply:
`if(something is nullptr)` is a bad idea, especially when it's avoidable. [`Fake<Construct>`](https://github.com/browsermt/marian-dev/issues/81#issuecomment-1079171876) is also a potential recipe for disaster.

Reply:
Nullptr is bad only when you don't handle it? You have two cases, either you have a bias or not?
https://github.com/marian-nmt/marian-dev/blob/23c36ec1a3c71cc75bc49fd3e39a4b1d8636589d/src/tensors/cpu/intgemm_interface.h#L73

Why would nullptr be a bad idea, c/c++ is full of places where you check if something is null?

Reply:
In your own provided example, intgemm does not accept `nullptr` in its signature. There's a bias that is set to `nullptr` while it could have been `true`/`false` and `child[2]` retrieved in the line below. Such usage of `nullptr` leads to checks everywhere and more error. See https://github.com/browsermt/marian-dev/issues/81 where someone passed a `nullptr` as an argument. https://github.com/browsermt/marian-dev/pull/82 is not okay as a fix, it's multiply without bias as a primitive (WASM Intrinsic) that is less prone to error, without `nullptr` in the signature.

https://www.youtube.com/watch?v=ybrQvs4x0Ps

Reply:
Not going to argue with you here. A default parameter `bias=nullptr` is a very obvious tell tale of what can and cannot be null. Std optional is not that much better.

Avoiding nullptr for control flow is good in general but not at the cost of 100% code duplication and / or preprocessor hacks.

Someone else wrote bad code is not an excuse for you to write duplicated code when a future code reviewer would scratch their head and wonder why it's duplicated.

Reply:
> `if(something is nullptr)` is a bad idea, especially when it's avoidable. [`Fake<Construct>`](https://github.com/browsermt/marian-dev/issues/81#issuecomment-1079171876) is also a potential recipe for disaster.

`nullptr` is literally how ruy implements optional bias terms.  https://github.com/google/ruy/blob/38a9266b832767a3f535a74a9e0cf39f7892e594/ruy/mul_params.h#L225

Reply:
These callbacks have very different semantics from intgemm.  In intgemm they're inline for every element there is a callback.  Here they are done over the full matrix.  I get what you're doing though it's confusing to use the same name for what are not drop-in replacements.